File: day02/day02-B.py
#!/usr/bin/env python3

import sys

# Functions

def process_commands(commands):
    position, depth, aim = 0, 0, 0

    for direction, magnitude in commands:
        magnitude = int(magnitude)

        if direction == 'down':
            aim += magnitude
        elif direction == 'up':
            aim -= magnitude
        elif direction == 'forward':
            position += magnitude
            depth    += aim * magnitude

    return position, depth

# Main Execution

def main():
    commands        = [line.split() for line in sys.stdin]
    position, depth = process_commands(commands)
    print(position * depth)

if __name__ == '__main__':
    main()

File: day02/day02-B.js
#!/usr/bin/env gjs

// Imports

imports.searchPath.unshift('../common');
const IO = imports.io;

// Functions

function process_commands(commands) {
    let [position, depth, aim] = [0, 0, 0];

    commands.forEach(command => {
    	let [direction, magnitude] = [command[0], parseInt(command[1])];

	if (direction == 'down') {
    	    aim += magnitude;
	} else if (direction == 'up') {
    	    aim -= magnitude;
	} else if (direction == 'forward') {
    	    position += magnitude;
    	    depth    += aim * magnitude;
	}
    });

    return position * depth;
}

// Main Execution

let commands = IO.readlines(s => s.split(' '));
let result   = process_commands(commands);
IO.print(result);
